### ISSUE TITLE
Mixed up language part information from different records

### ISSUE THREAD

**Issue by osma** [lcnetdev/marc2bibframe2#7]
Language information from the `041` fields can be mixed up when converting multiple records in the same batch.

For example, I have two records, with these 041 fields:
```xml
<!-- first record: translation from English to Finnish -->
    <marc:datafield tag="041" ind1="1" ind2=" ">
      <marc:subfield code="a">fin</marc:subfield>
      <marc:subfield code="h">eng</marc:subfield>
    </marc:datafield>

<!-- second record: main language is English, with a summary in Finnish -->
    <marc:datafield tag="041" ind1="0" ind2=" ">
      <marc:subfield code="a">eng</marc:subfield>
      <marc:subfield code="b">fin</marc:subfield>
    </marc:datafield>
```

The full records are [here](https://gist.github.com/osma/8fa1dc7c05cc53927722353f72a3e273).

When converting these, the language information is represented like this in the BIBFRAME2 RDF/XML output:
```xml
<!-- Work based on first record: translation from English to Finnish -->
    <bf:language>
      <bf:Language rdf:about="http://id.loc.gov/vocabulary/languages/fin">
        <bf:source>
          <bf:Source rdf:about="http://id.loc.gov/vocabulary/languages"/>
        </bf:source>
      </bf:Language>
    </bf:language>
    <bf:language>
      <bf:Language rdf:about="http://id.loc.gov/vocabulary/languages/eng">
        <bf:part>original</bf:part>
        <bf:source>
          <bf:Source rdf:about="http://id.loc.gov/vocabulary/languages"/>
        </bf:source>
      </bf:Language>
    </bf:language>

<!-- Work based on second record: main language is English, with a summary in Finnish -->
    <bf:language>
      <bf:Language rdf:about="http://id.loc.gov/vocabulary/languages/eng">
        <bf:source>
          <bf:Source rdf:about="http://id.loc.gov/vocabulary/languages"/>
        </bf:source>
      </bf:Language>
    </bf:language>
    <bf:language>
      <bf:Language rdf:about="http://id.loc.gov/vocabulary/languages/fin">
        <bf:part>summary</bf:part>
        <bf:source>
          <bf:Source rdf:about="http://id.loc.gov/vocabulary/languages"/>
        </bf:source>
      </bf:Language>
    </bf:language>
```

When this is parsed and interpreted as RDF, we get these triples (here in Turtle syntax):
```
# Work based on first record: translation from English to Finnish
fennica:006123458#Work
    bf:language [ a bf:Language ;
            rdfs:label "suomi" ],
        <http://id.loc.gov/vocabulary/languages/eng>,
        <http://id.loc.gov/vocabulary/languages/fin> ;

# Work based on second record: main language is English, with a summary in Finnish
fennica:006089019#Work
    bf:language <http://id.loc.gov/vocabulary/languages/eng>,
        <http://id.loc.gov/vocabulary/languages/fin> ;

# language entities referred to by both the above Works
<http://id.loc.gov/vocabulary/languages/eng> a bf:Language ;
    bf:part "original" ;
    bf:source <http://id.loc.gov/vocabulary/languages> .

<http://id.loc.gov/vocabulary/languages/fin> a bf:Language ;
    bf:part "summary" ;
    bf:source <http://id.loc.gov/vocabulary/languages> .
```

The problem is that now it is impossible to know, for each Work, which of the languages are "original" (or not) and which are "summary" (or not). It looks like both records had English as "original" language and Finnish as "summary" language, which is not true.

The languages are represented using URIs from the LOC languages vocabulary, and the `bf:part` properties are given to those language entities directly. There should probably be some indirection mechanism, so that Works are not directly assigned LOC language URIs but instead some intermediate resource, something like this:

```
:work bf:language [
  a bf:Language ;
  bf:part "original",
  ex:language <http://id.loc.gov/vocabulary/languages/eng>
] .
```

where `ex:language` would be replaced with some suitable property from the `bf:` or `bflc:` namespace.

**Comment by osma** [comment]
Thanks for the response @kirkhess.

A note about the *enhancement* label given to this issue: from my perspective, this is not a feature request, but a show-stopper bug! I cannot start using this converter until this is addressed (or I can find a way to work around it somehow). I need to be able to convert batches of records (thousands) at a time, and this issue means that I cannot know from the conversion result which languages are "real" and which are e.g. original languages of a translation, as the data about languages is all messed up.

**Comment by kirkhess** [comment]
I meant we need to analyze and update the specs for this particular problem
if we decide to change it, I'll see if I can come up with better labels.

**Comment by osma** [comment]
Any news? April came and went...

If you need help speccing this change, I'm happy to help!

**Comment by kirkhess** [comment]
Sorry about that, everything is in the backlog right now & we're going to
start another sprint next week.

-Kirk

**Comment by wafschneider** [comment]
AFAICT, this would be the only place that we use rdf:value indirection precisely this way. For converting $u in 5XX fields, we use an rdfs:label property with a xs:anyURI datatype. For converting URIs in $0 or $w, we create a bf:identifiedBy property with a bf:Identifier object that has an rdf:value property containing the URI.

My instinct would be that the latter approach (using a bf:identifiedBy property) is a little clearer. Other thoughts?

**Comment by wafschneider** [comment]
(by the way, @osma -- thanks a million for the pull request, and for pushing this issue forward!)

**Comment by osma** [comment]
@wafschneider I'm personally happy with any method of indirection that solves the original problem.

Actually, I'd prefer a solution where the type of language would be specified using a set of properties, e.g. `bf:originalLanguage`, `bf:summaryLanguage`, `bf:mainLanguage` etc. corresponding to the available MARC 041 subfields - there are 11 of these currently defined. These properties could all be subproperties of a generic `bf:language` property. Then the values itself could be language entities from the LC languages vocabulary, as they are already. There would be no need to assert any additional properties for the language entities, so no need for any sort of indirection.

As a variation of this, the original language could (also?) be asserted as the main language for the work entity that represents the original work.

**Comment by wafschneider** [comment]
I will bring your proposed solution to the maintainers of the ontology, maybe that's ultimately the best way to go about it. For now, we'll either implement your pull request as it is (perhaps adding a datatype), or add the additional layer of the bf:identifiedBy property to clarify, if the spec maintainers think that's a good idea. Thanks!

**Comment by osma** [comment]
Here are some further arguments for the subproperty pattern I proposed above:

1. It's simpler and reduces the number of triples and entities on the instance level, without losing any information.
2. It's easier to query using SPARQL. For example, getting the summary language of a work is simply `{ ?work bf:summaryLanguage ?sl }`, whereas with indirection it would be something like `{ ?work bf:language ?l . ?l bf:part "summary" ; ?l bf:identifiedBy ?sl }`. The simpler query would most likely be faster too.
3. It's more semantic - the distinction between different types of languages in encoded using RDF properties instead of literals.
4. It's language-agnostic and thus more suitable for internationalization. Labels can be defined for the new properties (like any properties in BIBFRAME) in any language using `rdfs:label` assertions, whereas the English language strings "original", "summary" etc. cannot easily be internationalized.

**Comment by wafschneider** [comment]
I've merged PR #10 into branch rc-1.2, hopefully merging to master and releasing this week. I added the additional layer of indirection using a bf:identifier property. I will raise a separate issue for the proposed ontology enhancement -- I think @osma's arguments are sound, and really worth further discussion.

**Comment by mnscholz** [comment]
Hi,
what is the idea behind the new conversion using bf:part and bf:identifier? It results in something like:
```
    <bf:language>
      <bf:Language>
        <bf:part>summary</bf:part>
        <bf:identifiedBy>
          <bf:Identifier>
            <rdf:value rdf:resource="http://id.loc.gov/vocabulary/languages/ger"/>
            <bf:source>
              <bf:Source rdf:about="http://id.loc.gov/vocabulary/languages"/>
            </bf:source>
          </bf:Identifier>
        </bf:identifiedBy>
      </bf:Language>
    </bf:language>
```
I think this conflicts with 
a) with the definition of `bf:Language` as representing (natural) languages that cannot have parts like a "summary". Or do I misunderstand the concept behind bf:Language?
b) with the use in other fields, where the langauge URI (e.g. <http://id.loc.gov/vocabulary/languages/ger>) is regarded to be a `bf:Language`, not a `bf:Identifier`.

**Comment by kirkhess** [comment]
We actually have this particular problem in other areas (relationships, contributions), so we're working on a new pattern where we create an intermediate resource using a new property bflc:target (which was basically Osma's suggestion above).  

:work bf:language [
  a bf:Language ;
  bf:part "summary",
  bflc:target <http://id.loc.gov/vocabulary/languages/eng>
] .

This is similar to the earlier attempt with bf:identifiedBy/bf:Identifier but it fixes the part problem.

**Comment by mnscholz** [comment]
Again, I think that the bnode in your example does not stand for a language as such (English), but rather for a part of the work or a part's "language aspect". 
This contrasts with the skos:definition of bf:Language and bf:language. The definition of the latter says: 
> Language associated with a resource or its parts.
(from http://id.loc.gov/ontologies/bibframe.rdf)

Which interpretation of bf:Language is right?
